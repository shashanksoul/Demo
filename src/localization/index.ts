import en from './translations/en.json';
import ar from './translations/ar.json';
import hi from './translations/hi.json';
import i18next from 'i18next';
import {initReactI18next} from 'react-i18next';

const languageResources = {
  en: {translation: en},
  ar: {translation: ar},
  hi: {translation: hi},
};

i18next.use(initReactI18next).init({
  compatibilityJSON: 'v3',
  lng: 'en',
  fallbackLng: 'en',
  resources: languageResources,
});

export default i18next;
