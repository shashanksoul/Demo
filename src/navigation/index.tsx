import {NavigationContainer} from '@react-navigation/native';
import {createNativeStackNavigator} from '@react-navigation/native-stack';
import React from 'react';
import {HomeScreen, LoginScreen} from '../screens';
import {RootStackParamList} from './types';

const Stack = createNativeStackNavigator<RootStackParamList>();

const RootNavigation: React.FC = () => {
  return (
    <NavigationContainer>
      <Stack.Navigator>
        <Stack.Screen
          options={{headerShown: false}}
          name="Login"
          component={LoginScreen}
        />
        <Stack.Screen name="Home" component={HomeScreen} />
      </Stack.Navigator>
    </NavigationContainer>
  );
};

export default RootNavigation;
