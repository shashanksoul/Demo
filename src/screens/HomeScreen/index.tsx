import React, {useEffect, useState} from 'react';
import {FlatList, Text, View} from 'react-native';
import MovieItem from '../../components/MovieItem';
import style from './styles';
import {useDispatch, useSelector} from 'react-redux';
import {MovieThunks} from '../../data/redux';
import {AppState} from '../../data/redux/types';
import ProgressBar from '../../components/ProgressBar';

const HomeScreen: React.FC = () => {
  const [page, setPage] = useState(1);
  const {isLoading, movies, error, currentPage, totalPages} = useSelector(
    (state: AppState) => ({
      isLoading: state.isLoading,
      movies: state.movies,
      error: state.error,
      currentPage: state.currentMoviePage,
      totalPages: state.totalPages,
    }),
  );
  const dispatch: any = useDispatch();

  useEffect(() => {
    dispatch(MovieThunks.getMovies(page));
  }, [dispatch, page]);

  return (
    <View style={style.container}>
      <Text>{error}</Text>
      <FlatList
        numColumns={2}
        onEndReachedThreshold={0.3}
        onEndReached={() => {
          if (currentPage && totalPages && currentPage <= totalPages) {
            setPage(prevPage => prevPage + 1);
          }
        }}
        data={movies}
        keyExtractor={(item, index) => `${item.title},${index}`}
        renderItem={({item}) => <MovieItem movieDetail={item} />}
      />
      {isLoading && <ProgressBar />}
    </View>
  );
};

export default HomeScreen;
