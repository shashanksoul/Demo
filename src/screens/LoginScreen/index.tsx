import React, {useState} from 'react';
import {Image, Text, TextInput, TouchableOpacity, View} from 'react-native';
import {style} from './styles';
import {isValidEmail, isValidPassword} from '../../utils/isValid';
import {LoginScreenProps} from './types';

const LoginScreen: React.FC<LoginScreenProps> = ({navigation}) => {
  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');

  const isLoginEnabled = () => isValidEmail(email) && isValidPassword(password);
  return (
    <View style={style.container}>
      <Image
        style={style.backgroundImage}
        source={require('../../assets/images/home-bg.jpg')}
      />
      <Text style={style.title}>Demo App</Text>
      <TextInput
        testID="email-input"
        value={email}
        onChangeText={setEmail}
        keyboardType="email-address"
        style={style.input}
        placeholder="User name"
      />
      <TextInput
        testID="password-input"
        value={password}
        onChangeText={setPassword}
        keyboardType="visible-password"
        style={style.input}
        placeholder="Password"
      />
      <TouchableOpacity
        testID="login-button"
        disabled={!isLoginEnabled()}
        activeOpacity={0.8}
        style={
          isLoginEnabled() ? style.button : [style.button, style.buttonDisabled]
        }
        onPress={() => navigation.replace('Home')}>
        <Text
          style={
            isLoginEnabled()
              ? style.buttonText
              : [style.buttonText, style.buttonDisabledText]
          }>
          Login
        </Text>
      </TouchableOpacity>
    </View>
  );
};

export default LoginScreen;
