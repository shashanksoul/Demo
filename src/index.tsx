import React, {useEffect} from 'react';
import RootNavigation from './navigation';
import {applyMiddleware, combineReducers, createStore} from 'redux';
import {Provider} from 'react-redux';
import {thunk} from 'redux-thunk';
import {authReducer, movieReducer} from './data/redux';
import './localization';
import {updateSavedLng} from './utils/language';
import {PersistGate} from 'redux-persist/integration/react';
import {persistStore, persistReducer} from 'redux-persist';
import storage from '@react-native-async-storage/async-storage';

const persistConfig = {
  key: 'root',
  storage: storage,
  whitelist: ['authReducer'],
};

const rootReducer = combineReducers({
  movieReducer,
  authReducer,
});

const persistedReducer = persistReducer(persistConfig, rootReducer);

const middleWares = [thunk];

const store = createStore(persistedReducer, applyMiddleware(...middleWares));
const persistor = persistStore(store);
const App = () => {
  useEffect(() => {
    (async () => await updateSavedLng())();
  }, []);
  return (
    <Provider store={store}>
      <PersistGate loading={null} persistor={persistor}>
        <RootNavigation />
      </PersistGate>
    </Provider>
  );
};

export default App;
