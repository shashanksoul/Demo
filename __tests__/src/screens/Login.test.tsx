import {fireEvent, render} from '@testing-library/react-native';
import {LoginScreen} from '../../../src/screens';
import React from 'react';

const mockNavigation: any = {
  replace: jest.fn(),
};
const mockRoute: any = jest.fn();

describe('Login Screen', () => {
  it('renders correctly', () => {
    const loginUI = render(
      <LoginScreen navigation={mockNavigation} route={mockRoute} />,
    );
    const button = loginUI.getByTestId('login-button');
    fireEvent(button, 'press');
    expect(mockNavigation.replace).not.toHaveBeenCalled();
    expect(loginUI).toMatchSnapshot();
  });

  it('when username not valid', () => {
    const {getByTestId} = render(
      <LoginScreen navigation={mockNavigation} route={mockRoute} />,
    );
    const passwordInput = getByTestId('password-input');
    const emailInput = getByTestId('email-input');
    fireEvent.changeText(passwordInput, 'Kuy@1298h9');
    fireEvent.changeText(emailInput, 'skumargmail.com');
    const button = getByTestId('login-button');
    fireEvent(button, 'press');
    expect(passwordInput.props.value).toBe('Kuy@1298h9');
    expect(emailInput.props.value).toBe('skumargmail.com');
    expect(mockNavigation.replace).not.toHaveBeenCalled();
  });

  it('when password not valid', () => {
    const {getByTestId} = render(
      <LoginScreen navigation={mockNavigation} route={mockRoute} />,
    );
    const passwordInput = getByTestId('password-input');
    const emailInput = getByTestId('email-input');
    fireEvent.changeText(passwordInput, 'asf345fa');
    fireEvent.changeText(emailInput, 'skumar@gmail.com');
    const button = getByTestId('login-button');
    fireEvent(button, 'press');
    expect(passwordInput.props.value).toBe('asf345fa');
    expect(emailInput.props.value).toBe('skumar@gmail.com');
    expect(mockNavigation.replace).not.toHaveBeenCalled();
  });

  it('when email & password valid', () => {
    const {getByTestId} = render(
      <LoginScreen navigation={mockNavigation} route={mockRoute} />,
    );
    const passwordInput = getByTestId('password-input');
    const emailInput = getByTestId('email-input');
    fireEvent.changeText(passwordInput, 'Kuy@1298h9');
    fireEvent.changeText(emailInput, 'skumar@gmail.com');
    const button = getByTestId('login-button');
    fireEvent(button, 'press');
    expect(mockNavigation.replace).toHaveBeenCalledWith('Home');
  });
});
