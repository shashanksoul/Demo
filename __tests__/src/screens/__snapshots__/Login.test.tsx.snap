// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Screen renders correctly 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "backgroundColor": "black",
      "flex": 1,
      "justifyContent": "center",
      "padding": 20,
    }
  }
>
  <Image
    source={
      {
        "testUri": "../../../src/assets/images/home-bg.jpg",
      }
    }
    style={
      {
        "position": "absolute",
      }
    }
  />
  <Text
    style={
      {
        "color": "white",
        "fontSize": 40,
        "fontWeight": "bold",
        "marginBottom": 50,
      }
    }
  >
    Demo App
  </Text>
  <TextInput
    keyboardType="email-address"
    onChangeText={[Function]}
    placeholder="User name"
    style={
      {
        "backgroundColor": "#FFFFFF80",
        "borderRadius": 20,
        "margin": 10,
        "paddingHorizontal": 20,
        "paddingVertical": 10,
        "width": "100%",
      }
    }
    testID="email-input"
    value=""
  />
  <TextInput
    keyboardType="visible-password"
    onChangeText={[Function]}
    placeholder="Password"
    style={
      {
        "backgroundColor": "#FFFFFF80",
        "borderRadius": 20,
        "margin": 10,
        "paddingHorizontal": 20,
        "paddingVertical": 10,
        "width": "100%",
      }
    }
    testID="password-input"
    value=""
  />
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": true,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#80808050",
        "borderRadius": 30,
        "justifyContent": "center",
        "marginTop": 30,
        "opacity": 1,
        "paddingVertical": 20,
        "width": "60%",
      }
    }
    testID="login-button"
  >
    <Text
      style={
        [
          {
            "color": "white",
          },
          {
            "color": "black",
          },
        ]
      }
    >
      Login
    </Text>
  </View>
</View>
`;
